content/derived-datatypes.rst

```diff
@@ -11,47 +11,461 @@ Derived datatypes
-   - Learn to send and receive composite messages with |term-MPI_Pack| and |term-MPI_Unpack|
+   - Understand how MPI handles datatypes.
+   - Learn to send and receive composite messages
    - Learn how to represent homogeneous collections as MPI messages. |term-MPI_Type_contiguous|, |term-MPI_Type_vector|, |term-MPI_Type_indexed|
    - Learn how to represent your own derived datatypes as MPI messages with |term-MPI_Type_create_struct| and |term-MPI_Type_commit|
-   -  MAYBE: Learn how to define combination operations on your datatypes: |term-MPI_Op_create| and |term-MPI_Op_free|
 
 
-Topic introduction here
+The ability to define custom datatypes is one of the hallmarks of a modern
+programming language, since it allows programmers to structure their code in a
+way that enhances readability and maintainability.
+How can this be done in MPI? Recall that MPI is a standard describing a library
+to enable parallel programming in the message passing model.
+
+MPI supports many of the basic datatypes recognized by the C standard.
+
+
+.. csv-table:: Basic datatypes in MPI and in the C standard. For a comprehensive
+               explanation of the types defined in the C language, you can
+               consult `this reference
+               <https://en.cppreference.com/w/c/language/type>`_.
+   :widths: auto
+   :align: center
+   :header-rows: 1
+   :delim: ;
+
+   MPI ; C
+   ``MPI_CHAR`` ; ``signed char``
+   ``MPI_FLOAT`` ; ``float``
+   ``MPI_DOUBLE`` ; ``double``
+   ``MPI_LONG_DOUBLE`` ; ``long double``
+   ``MPI_WCHAR`` ; ``wchar_t``
+   ``MPI_SHORT`` ; ``short``
+   ``MPI_INT`` ; ``int``
+   ``MPI_LONG`` ; ``long``
+   ``MPI_LONG_LONG_INT`` ; ``long long``
+   ``MPI_SIGNED_CHAR`` ; ``signed char``
+   ``MPI_UNSIGNED_CHAR`` ; ``unsigned char``
+   ``MPI_UNSIGNED_SHORT`` ; ``unsigned short``
+   ``MPI_UNSIGNED`` ; ``unsigned int``
+   ``MPI_UNSIGNED_LONG`` ; ``unsigned long``
+   ``MPI_UNSIGNED_LONG_LONG`` ; ``unsigned long long``
+   ``MPI_C_COMPLEX`` ; ``float _Complex``
+   ``MPI_C_DOUBLE_COMPLEX`` ; ``double _Complex``
+   ``MPI_C_LONG_DOUBLE_COMPLEX`` ; ``long double _Complex``
+   ``MPI_PACKED`` ;
+   ``MPI_BYTE`` ;
+
+
+In the C language, types are **primitive** constructs: they
+are *defined* by the standard and *enforced* by the compiler.
+The MPI types are instead **variants** in the ``MPI_Datatype`` enumeration: they
+appear as the **same** type to the compiler.
+This is a fundamental difference which influences the way custom datatypes are handled.
+
+In the C language, you would declare a ``struct`` such as the following:
 
-You really want to browse this page alongside the source of it, to see
-how this is implemented.  See the links at the to right of the page.
 
+.. code-block:: c
+
+   struct Pair {
+    int first;
+    char second;
+   };
+
+``Pair`` is a new type. From the compiler's point of view, it has status on par
+with the fundamental datatypes introduced above. The C standard mandates *how
```

Comment:
```suggestion
with the fundamental datatypes introduced above. The C standard makes requirements on *how
```

since details like alignment requirements are implementation-specific. Memory layout of composite types in C is not universally portable. (If it was, then MPI would not need to jump through hoops at run time to learn about the layout!)

Reply:
Indeed! :smile_cat: